### ISSUE TITLE
Mistral-7b model in rust candle example is slower than python Mistral-7b

### ISSUE THREAD

**Issue by dlfrnaos19** [huggingface/candle#1003]
I expected the Rust version to be faster, but it is actually slower than the Python version, 

obviously, I tested a few times more, so I did warming up.

please give me advice for some accurate benchmark so I can share it after testing.

cpu: AMD Ryzen 9 5950x 16-Core Processor
Ram: 128GB
GPU: RTX 3090 24GB

rust docker image -> rust official docker image,  cuda 12.2 from host to  container
python docker image -> datascience-notebook,  cuda 12.2 from host to  container 

The candle library was tested by copying the GitHub source.

python main library versions
torch 2.0.1 + cu118
transformers 4.34.0.dev0
accelerate 0.23.0

python

rust example link : https://github.com/huggingface/candle/tree/main/candle-examples/examples/mistral

rust command
cargo run --example mistral --release --features cuda -- --prompt 'Write fibonacci code in golang' --sample-len 300

result
**35.94 token/s**

python code
```
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
import time
device = "cuda" # the device to load the model onto

model_path = "mistralai/Mistral-7B-Instruct-v0.1"

# torch.float32 OOM
model = AutoModelForCausalLM.from_pretrained(model_path, torch_dtype=torch.float16)
tokenizer = AutoTokenizer.from_pretrained(model_path)
model.to(device)

start = time.time()
message = "<s>[INST] write fibonacci code in golang [/INST]"
input_ids = tokenizer(message,return_tensors="pt")["input_ids"].to(device)
generated_ids = model.generate(input_ids, max_length=300, do_sample=True)
decoded = tokenizer.batch_decode(generated_ids)
end = time.time() - start
token_per_second = generated_ids[0][17:].shape[0] / end

print("result",decoded[0])
print("token length", generated_ids[0][17:].shape[0])
print("token per second", token_per_second)
```

result
**42.58 token/s**

**Comment by LaurentMazare** [comment]
Thanks for reporting this. One difference is that the rust version does not use very optimized ops at the moment. I just merged support for flash-attn, this used to make a large difference on the original llama2 model but somehow here it did not make much difference in my testing. Could you still give it a spin, e.g. by running:
```bash
cargo run --example mistral --release --features cuda,flash-attn -- --prompt 'Write helloworld code in Rust' --sample-len 250 --use-flash-attn
```
The downside of flash-attn is some long compilation time. You can set the `CANDLE_FLASH_ATTN_BUILD_DIR` environment variable to something like `/home/user/.candle` to ensures that the compilation artifacts are properly cached.

**Comment by dlfrnaos19** [comment]
Thank you for your quick response.
I ran it again with the command you provided, an error occurred during the flash-attention build.

error head
```
error: failed to run custom build command for `candle-flash-attn v0.2.3 (/root/work/candle/candle-flash-attn)`

Caused by:
  process didn't exit successfully: `/root/work/candle/target/release/build/candle-flash-attn-1f089548c379ba67/build-script-build` (exit status: 1)
```

error tail
```
Error: nvcc error while executing compiling: "nvcc" "-std=c++17" "--gpu-architecture=sm_86" "-c" "-o" "/root/work/candle/target/release/build/candle-flash-attn-ef8d50c18aa1cabf/out/flash_api.o" "--default-stream" "per-thread" "-Icutlass/include" "--expt-relaxed-constexpr" "--verbose" "kernels/flash_api.cu"
```

do you think I have to change some Cuda version or Rust version?
my Rust version is rustc 1.72.1

**Comment by LaurentMazare** [comment]
It's hard to tell without the full error but you may want to check the [FAQ](https://github.com/huggingface/candle#missing-cutecutlass-headers-when-compiling-flash-attn) as a common issue is not initializing the git submodule for cutlass.

**Comment by dlfrnaos19** [comment]
Thank you for your help
The issue was with cutlass, and it has been resolved.
I ran the code number 5 and here are the results.

35.96, 35.98, 35.94, 36.13, 36.16

**Comment by LaurentMazare** [comment]
Thanks for reporting this, looks a bit similar to what I observed i.e. no real upside with flash-attn which is quite surprising to me. The time is probably wasted on other ops, I guess we will need to make a bit of profiling here to get to the bottom of this.

**Comment by dlfrnaos19** [comment]
I agree that profiling could help identify where the time is being spent and potentially reveal opportunities for optimization. I'll continue to monitor the performance on my end as well. If I come across any findings that could be helpful, I'll be sure to share them. Thanks again for your support and guidance.

**Comment by guoqingbao** [comment]
> Thanks for reporting this, looks a bit similar to what I observed i.e. no real upside with flash-attn which is quite surprising to me. The time is probably wasted on other ops, I guess we will need to make a bit of profiling here to get to the bottom of this.

I have profiled the candle ops before (we have a project to porting candle to a custom DSA), the current version of rotary embedding, concat (kv concat), and layernorm need to be improved (e. g.  fused kernels). From our experience,  we achieved over 200% performance improvements by making fused version of the above kernels on DSA. Another potential improvement is the inplace op. Every op in the current version of candle need to return a copy or allocate a output buffer to the subsequent ops which is not efficient. Actually, most of the unary ops (including softmax,  layernorm) can be calculated inplace. There are also redundant memset (i. e., fill the buffer with values and then copy values to the buffer) in concat, reshape, contigous ops. We proposed a git merge request #1680 to fix the inplace and redundant memset issues, we are also happy to help write and test fused rotary embedding, layernorm, softmax for candle.
